docs/proposals/deployment.md

```diff
@@ -0,0 +1,269 @@
+  // selected by this will be scaled down.
+  Selector map[string]string
+
+  // Describes the pods that will be created.
+  Template *PodTemplateSpec
+
+  // The deployment strategy to use to replace existing pods with new ones.
+  Strategy DeploymentStrategy
+
+  // Key of the selector that is added to existing RCs (and label key that is
+  // added to its pods) to prevent the existing RCs to select new pods (and old
+  // pods being selected by new RC).
+  // Users can set this to an empty string to indicate that the system should
+  // not add any selector and label. If unspecified, system uses
+  // "deployment.kubernetes.io/podTemplateHash".
+  // Value of this key is hash of DeploymentSpec.PodTemplateSpec.
+  UniqueLabelKey *string
+}
+
+type DeploymentStrategy struct {
+  // Type of deployment. Can be "Recreate" or "RollingUpdate".
+  Type DeploymentType
+
+  // TODO: Update this to follow our convention for oneOf, whatever we decide it
+  // to be.
+  // Rolling update config params. Present only if DeploymentType =
+  // RollingUpdate.
+  RollingUpdate *RollingUpdateDeploymentSpec
+}
+
+type DeploymentType string
+
+const (
+  // Kill all existing pods before creating new ones.
+  DeploymentRecreate DeploymentType = "Recreate"
+
+  // Replace the old RCs by new one using rolling update i.e gradually scale down the old RCs and scale up the new one.
+  DeploymentRollingUpdate DeploymentType = "RollingUpdate"
+)
+
+// Spec to control the desired behavior of rolling update.
+type RollingUpdateDeploymentSpec struct {
+  // The maximum number of pods that can be unavailable during the update.
+  // Value can be an absolute number (ex: 5) or a percentage of total pods at the start of update (ex: 10%).
+  // Absolute number is calculated from percentage by rounding up.
+  // This can not be 0 if MaxSurge is 0.
+  // By default, a fixed value of 1 is used.
+  // Example: when this is set to 30%, the old RC can be scaled down by 30%
+  // immediately when the rolling update starts. Once new pods are ready, old RC
+  // can be scaled down further, followed by scaling up the new RC, ensuring
+  // that at least 70% of original number of pods are available at all times
+  // during the update.
+  MaxUnavailable IntOrString
+
+  // The maximum number of pods that can be scheduled above the original number of
+  // pods.
+  // Value can be an absolute number (ex: 5) or a percentage of total pods at
+  // the start of the update (ex: 10%). This can not be 0 if MaxUnavailable is 0.
+  // Absolute number is calculated from percentage by rounding up.
+  // By default, a value of 1 is used.
+  // Example: when this is set to 30%, the new RC can be scaled up by 30%
+  // immediately when the rolling update starts. Once old pods have been killed,
+  // new RC can be scaled up further, ensuring that total number of pods running
+  // at any time during the update is atmost 130% of original pods.
+  MaxSurge IntOrString
```

Comment:
On Aug 14, 2015, at 6:43 PM, James Eady <notifications@github.com> wrote:

In docs/proposals/deployment.md
<https://github.com/kubernetes/kubernetes/pull/12236#discussion_r37126196>:

> +  // can be scaled down further, followed by scaling up the new RC, ensuring
> +  // that at least 70% of original number of pods are available at all times
> +  // during the update.
> +  MaxUnavailable IntOrString
> +
> +  // The maximum number of pods that can be scheduled above the original number of
> +  // pods.
> +  // Value can be an absolute number (ex: 5) or a percentage of total pods at
> +  // the start of the update (ex: 10%). This can not be 0 if MaxUnavailable is 0.
> +  // Absolute number is calculated from percentage by rounding up.
> +  // By default, a value of 1 is used.
> +  // Example: when this is set to 30%, the new RC can be scaled up by 30%
> +  // immediately when the rolling update starts. Once old pods have been killed,
> +  // new RC can be scaled up further, ensuring that total number of pods running
> +  // at any time during the update is atmost 130% of original pods.
> +  MaxSurge IntOrString

I find maxUnavailable and maxSurge to be somewhat confusing. Doesn't this
mean that, for a given update that doesn't change the size, the number of
pods running at any given time can vary by (maxUnavailable + maxSurge),
entirely depending on the updater implementation?

Is there a demand to have separate knobs for both of these bounds? It seems
like they could probably both be controlled by a single number for
simplicity. Or, given a more well-defined updater, you might only ever need
a single one of these knobs. For example, if the updater guarantees that
the new pod is running before the old pod is taken down, you'd only need
maxSurge.


To have both sides, yes.  We need to support deployments where the existing
deployment is consuming all physical resources, which requires the caller
to specify a lower bound.  See some of the upthread but we decided to keep
them split so as to better reflect their impact, as well as to describe the
policy more clearly to users.  The previous was a negative or positive
percentage that was also confusing.


—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12236/files#r37126196>.